core/embed/rust/src/ui/shape/cache/zlib_cache.rs

```diff
@@ -0,0 +1,173 @@
+            .init(UnsafeCell::new([0; UZLIB_WINDOW_SIZE]));
+
+        Some(Self {
+            zdata: &[],
+            offset: 0,
+            dc: None,
+            window,
+        })
+    }
+
+    /// May be called with zdata == &[] to make the slot free
+    fn reset(&mut self, zdata: &'a [u8]) {
+        // Drop the existing decompression context holding
+        // a mutable reference to window buffer
+        self.dc = None;
+
+        if !zdata.is_empty() {
+            // Now there's nobody else holding any reference to our window
+            // so we can get mutable reference and pass it to a new
+            // instance of the decompression context
+            let window = unsafe { &mut *self.window.get() };
+
+            self.dc = Some(UzlibContext::new(zdata, Some(window)));
+        }
+
+        self.offset = 0;
+        self.zdata = zdata;
+    }
+
+    fn uncompress(&mut self, dest_buf: &mut [u8]) -> Result<bool, ()> {
+        if let Some(dc) = self.dc.as_mut() {
+            match dc.uncompress(dest_buf) {
+                Ok(done) => {
+                    if done {
+                        self.reset(&[]);
+                    } else {
+                        self.offset += dest_buf.len();
+                    }
+                    Ok(done)
+                }
+                Err(e) => Err(e),
+            }
+        } else {
+            Err(())
+        }
+    }
+
+    fn skip(&mut self, nbytes: usize) -> Result<bool, ()> {
+        if let Some(dc) = self.dc.as_mut() {
+            match dc.skip(nbytes) {
+                Ok(done) => {
+                    if done {
+                        self.reset(&[]);
+                    } else {
+                        self.offset += nbytes;
+                    }
+
+                    Ok(done)
+                }
+                Err(e) => Err(e),
+            }
+        } else {
+            Err(())
+        }
+    }
+
+    fn is_for(&self, zdata: &[u8], offset: usize) -> bool {
+        self.zdata == zdata && self.offset == offset
+    }
+}
+
+pub struct ZlibCache<'a> {
+    slots: FixedVec<'a, ZlibCacheSlot<'a>>,
+}
+
+impl<'a> ZlibCache<'a> {
+    pub fn new<'alloc: 'a, T>(bump: &'alloc T, slot_count: usize) -> Option<Self>
+    where
+        T: LocalAllocLeakExt<'alloc>,
+    {
+        let mut cache = Self {
+            slots: bump.fixed_vec(slot_count)?,
+        };
+
+        for _ in 0..cache.slots.capacity() {
+            unwrap!(cache.slots.push(ZlibCacheSlot::new(bump)?)); // should never fail
+        }
+
+        Some(cache)
+    }
+
+    fn select_slot_for_reuse(&self) -> Result<usize, ()> {
+        if self.slots.capacity() > 0 {
+            let mut selected = 0;
+            for (i, slot) in self.slots.iter().enumerate() {
+                if slot.dc.is_none() {
+                    selected = i;
+                    break;
+                }
+            }
+            Ok(selected)
+        } else {
+            Err(())
+        }
+    }
```

Comment:
Yes, I was already aware of this issue. I've made a slight improvement to the approach - now we select the slot that has "performed the least amount of work", see 6a9e1b53448e7d9d7f591b6fa7ce24804f977bc0. 

I have decided to leave the code that checks for an empty list unchanged. This is because it's still theoretically possible to create a `ZlibCache` with zero slots.

Reply:
> I have decided to leave the code that checks for an empty list unchanged. This is because it's still theoretically possible to create a `ZlibCache` with zero slots.

sure, but we should be rejecting that possibility in `new()`. or do you have a usecase?

Reply:
No, I don't have any use case right now. The idea was that you would be able to customize the drawing cache before drawing any screen - so the `DrawingCache` will allocate just what you need and can better utilize the assigned memory chunk. I don’t think this check is a problem. Do you?